File: src/main/java/com/braydenoneal/block/entity/ControllerBlockEntity.java
package com.braydenoneal.block.entity;

import com.braydenoneal.blang.Context;
import com.braydenoneal.blang.parser.Program;
import com.braydenoneal.blang.parser.Scope;
import com.braydenoneal.blang.parser.expression.value.Value;
import com.braydenoneal.block.CableBlock;
import com.mojang.serialization.Codec;
import net.fabricmc.fabric.api.screenhandler.v1.ExtendedScreenHandlerFactory;
import net.minecraft.block.Block;
import net.minecraft.block.BlockState;
import net.minecraft.block.entity.BlockEntity;
import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.entity.player.PlayerInventory;
import net.minecraft.nbt.NbtCompound;
import net.minecraft.registry.RegistryWrapper;
import net.minecraft.screen.ScreenHandler;
import net.minecraft.server.network.ServerPlayerEntity;
import net.minecraft.storage.ReadView;
import net.minecraft.storage.WriteView;
import net.minecraft.text.Text;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.math.Direction;
import net.minecraft.world.World;
import org.jetbrains.annotations.Nullable;

import java.util.*;

import static net.minecraft.world.RedstoneView.DIRECTIONS;

public class ControllerBlockEntity extends BlockEntity implements ExtendedScreenHandlerFactory<BlockPos> {
    private Program program;
    private Map<String, Value<?>> variables;
    private String source;

    public ControllerBlockEntity(BlockPos pos, BlockState state) {
        super(ModBlockEntities.CONTROLLER_BLOCK_ENTITY, pos, state);
        source = "name;";
        program = new Program(source, new Context(world, pos, this));
        variables = Map.of();
    }

    public String source() {
        return source;
    }

    public Program program() {
        return program;
    }

    public void setSource(String source) {
        this.source = source;
        program = new Program(source, new Context(world, pos, this));
        program.run();
        variables = program.topScope().variables();
        markDirty();
    }

    @Override
    protected void readData(ReadView view) {
        super.readData(view);
        variables = view.read("variables", Scope.VARIABLES_CODEC).orElse(Map.of());
        source = view.read("source", Codec.STRING).orElse("name;");

        program = new Program(source, new Context(world, pos, this));
        program.topScope().setVariables(variables);
    }

    @Override
    protected void writeData(WriteView view) {
        super.writeData(view);
        view.put("variables", Scope.VARIABLES_CODEC, variables);
        view.put("source", Codec.STRING, source);
    }

    @Override
    public NbtCompound toInitialChunkDataNbt(RegistryWrapper.WrapperLookup registryLookup) {
        return createNbt(registryLookup);
    }

    public static void tick(World world, BlockPos blockPos, BlockState ignoredBlockState, ControllerBlockEntity entity) {
        if (world.isReceivingRedstonePower(blockPos)) {
            entity.program.runMain();
            entity.variables = entity.program.topScope().variables();
            entity.markDirty();
        }
    }

    @Override
    public Text getDisplayName() {
        return Text.translatable(getCachedState().getBlock().getTranslationKey());
    }

    @Override
    public @Nullable ScreenHandler createMenu(int syncId, PlayerInventory playerInventory, PlayerEntity player) {
        return new ControllerScreenHandler(syncId, playerInventory, pos);
    }

    @Override
    public BlockPos getScreenOpeningData(ServerPlayerEntity player) {
        return pos;
    }

    public List<ControllerBlockEntity> getConnectedControllerBlockEntities() {
        Set<BlockPos> cables = new HashSet<>();
        List<BlockPos> networkBlocks = new ArrayList<>();

        Stack<BlockPos> stack = new Stack<>();
        stack.push(pos);

        while (!stack.isEmpty()) {
            BlockPos pos = stack.pop();

            for (Direction direction : DIRECTIONS) {
                BlockPos adjacentPos = pos.offset(direction);

                if (cables.contains(adjacentPos)) {
                    continue;
                }

                assert world != null;
                Block adjacentBlock = world.getBlockState(adjacentPos).getBlock();

                if (adjacentBlock instanceof CableBlock) {
                    stack.push(adjacentPos);
                    cables.add(adjacentPos);
                }

                if (networkBlocks.contains(adjacentPos)) {
                    continue;
                }

                BlockEntity adjacentBlockEntity = world.getBlockEntity(adjacentPos);

                if (adjacentBlockEntity instanceof ControllerBlockEntity) {
                    networkBlocks.add(adjacentPos);
                }
            }
        }

        List<ControllerBlockEntity> controllers = new ArrayList<>();

        for (BlockPos pos : networkBlocks) {
            BlockEntity adjacentBlockEntity = world.getBlockEntity(pos);

            if (adjacentBlockEntity instanceof ControllerBlockEntity controller) {
                controllers.add(controller);
            }
        }

        return controllers;
    }
}

File: src/main/kotlin/com/braydenoneal/block/entity/ControllerBlockEntity.kt
package com.braydenoneal.block.entity

import com.braydenoneal.blang.wrapper.BlogicProgram
import com.braydenoneal.blang.wrapper.Context
import com.braydenoneal.blang.wrapper.codec.Codecs
import com.braydenoneal.block.CableBlock
import com.mojang.serialization.Codec
import net.fabricmc.fabric.api.screenhandler.v1.ExtendedScreenHandlerFactory
import net.minecraft.block.BlockState
import net.minecraft.block.entity.BlockEntity
import net.minecraft.block.entity.LockableContainerBlockEntity
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.entity.player.PlayerInventory
import net.minecraft.nbt.NbtCompound
import net.minecraft.registry.RegistryWrapper.WrapperLookup
import net.minecraft.screen.ScreenHandler
import net.minecraft.server.network.ServerPlayerEntity
import net.minecraft.state.property.Properties
import net.minecraft.storage.ReadView
import net.minecraft.storage.WriteView
import net.minecraft.text.Text
import net.minecraft.util.math.BlockPos
import net.minecraft.util.math.Direction
import net.minecraft.world.RedstoneView
import net.minecraft.world.World
import parser.expression.value.Value
import java.util.*
import kotlin.jvm.optionals.getOrNull

class ControllerBlockEntity(pos: BlockPos, state: BlockState) : BlockEntity(ModBlockEntities.CONTROLLER_BLOCK_ENTITY, pos, state), ExtendedScreenHandlerFactory<BlockPos> {
    private var source: String = "name;"
    private var program: BlogicProgram = BlogicProgram(source, Context(pos, this))
    private var variables: MutableMap<String, Value<*>> = mutableMapOf()
    private var initializing = true

    fun source(): String {
        return source
    }

    fun program(): BlogicProgram {
        return program
    }

    fun setSource(source: String) {
        this.source = source

        if (!world!!.isClient) {
            program = BlogicProgram(source, Context(pos, this))
            initializing = true
            variables = program.topScope().variables()
        }

        markDirty()
    }

    override fun readData(view: ReadView) {
        super.readData(view)
        variables = view.read("variables", Codecs.VARIABLES_CODEC).getOrNull() ?: mutableMapOf()
        source = view.read("source", Codec.STRING).getOrNull() ?: "name;"

        program = BlogicProgram(source, Context(pos, this))
        program.topScope().setVariables(HashMap(variables))
    }

    override fun writeData(view: WriteView) {
        super.writeData(view)
        view.put("variables", Codecs.VARIABLES_CODEC, variables)
        view.put("source", Codec.STRING, source)
    }

    override fun toInitialChunkDataNbt(registryLookup: WrapperLookup): NbtCompound {
        return createNbt(registryLookup)
    }

    val facing: Direction get() = cachedState.get(Properties.FACING)

    override fun getDisplayName(): Text {
        return Text.translatable(cachedState.block.getTranslationKey())
    }

    override fun createMenu(syncId: Int, playerInventory: PlayerInventory, player: PlayerEntity): ScreenHandler {
        return ControllerScreenHandler(syncId, playerInventory, pos)
    }

    override fun getScreenOpeningData(player: ServerPlayerEntity): BlockPos {
        return pos
    }

    fun getConnectedControllerBlockEntities(): MutableList<ControllerBlockEntity> {
        val cables: MutableSet<BlockPos> = HashSet()
        val networkBlocks: MutableList<BlockPos> = ArrayList()

        val stack = Stack<BlockPos>()

        for (direction in RedstoneView.DIRECTIONS) {
            val adjacentPos = pos.offset(direction)

            val adjacentBlock = world!!.getBlockState(adjacentPos).block

            if (adjacentBlock is CableBlock) {
                stack.push(adjacentPos)
                cables.add(adjacentPos)
            }
        }

        while (!stack.isEmpty()) {
            val pos = stack.pop()

            for (direction in RedstoneView.DIRECTIONS) {
                val adjacentPos = pos.offset(direction)

                if (cables.contains(adjacentPos)) {
                    continue
                }

                val adjacentBlock = world!!.getBlockState(adjacentPos).block

                if (adjacentBlock is CableBlock) {
                    stack.push(adjacentPos)
                    cables.add(adjacentPos)
                }

                if (networkBlocks.contains(adjacentPos)) {
                    continue
                }

                val adjacentBlockEntity = world!!.getBlockEntity(adjacentPos)

                if (adjacentBlockEntity is ControllerBlockEntity) {
                    networkBlocks.add(adjacentPos)
                }
            }
        }

        val controllers: MutableList<ControllerBlockEntity> = ArrayList()

        for (pos in networkBlocks) {
            val adjacentBlockEntity = world!!.getBlockEntity(pos)

            if (adjacentBlockEntity is ControllerBlockEntity) {
                controllers.add(adjacentBlockEntity)
            }
        }

        return controllers
    }

    fun getConnectedContainers(): MutableList<LockableContainerBlockEntity> {
        val cables: MutableSet<BlockPos> = HashSet()
        val networkBlocks: MutableList<BlockPos> = ArrayList()

        val stack = Stack<BlockPos>()

        for (direction in RedstoneView.DIRECTIONS) {
            val adjacentPos = pos.offset(direction)

            val adjacentBlock = world!!.getBlockState(adjacentPos).block

            if (adjacentBlock is CableBlock) {
                stack.push(adjacentPos)
                cables.add(adjacentPos)
            }
        }

        while (!stack.isEmpty()) {
            val pos = stack.pop()

            for (direction in RedstoneView.DIRECTIONS) {
                val adjacentPos = pos.offset(direction)

                if (cables.contains(adjacentPos)) {
                    continue
                }

                val adjacentBlock = world!!.getBlockState(adjacentPos).block

                if (adjacentBlock is CableBlock) {
                    stack.push(adjacentPos)
                    cables.add(adjacentPos)
                }

                if (networkBlocks.contains(adjacentPos)) {
                    continue
                }

                val adjacentBlockEntity = world!!.getBlockEntity(adjacentPos)

                if (adjacentBlockEntity is LockableContainerBlockEntity) {
                    networkBlocks.add(adjacentPos)
                }
            }
        }

        val controllers: MutableList<LockableContainerBlockEntity> = ArrayList()

        for (pos in networkBlocks) {
            val adjacentBlockEntity = world!!.getBlockEntity(pos)

            if (adjacentBlockEntity is LockableContainerBlockEntity) {
                controllers.add(adjacentBlockEntity)
            }
        }

        return controllers
    }

    companion object {
        fun tick(world: World, blockPos: BlockPos, ignoredBlockState: BlockState, entity: ControllerBlockEntity) {
            if (entity.initializing) {
                val result = entity.program.tick()

                if (result != null) {
                    entity.initializing = false
                }

                return
            }

            if (world.isReceivingRedstonePower(blockPos)) {
                entity.program.runMain()
                entity.variables = entity.program.topScope().variables()
                entity.markDirty()
            }
        }
    }
}
